repo/blob/azure/azure_storage.go

```diff
@@ -0,0 +1,273 @@
+// Package azure implements Azure Blob Storage.
+package azure
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"io/ioutil"
+
+	"github.com/efarrer/iothrottler"
+	"github.com/pkg/errors"
+	"gocloud.dev/gcerrors"
+
+	"github.com/kopia/kopia/internal/retry"
+	"github.com/kopia/kopia/repo/blob"
+
+	"github.com/Azure/azure-storage-blob-go/azblob"
+	gblob "gocloud.dev/blob"
+	"gocloud.dev/blob/azureblob"
+)
+
+const (
+	azStorageType   = "azure"
+	writerChunkSize = 1 << 20
+)
+
+type azStorage struct {
+	Options
+
+	ctx context.Context
+
+	bucket *gblob.Bucket
+
+	downloadThrottler *iothrottler.IOThrottlerPool
+	uploadThrottler   *iothrottler.IOThrottlerPool
+}
+
+func (az *azStorage) GetBlob(ctx context.Context, b blob.ID, offset, length int64) ([]byte, error) {
+	attempt := func() (interface{}, error) {
+		// TODO: Support chunked download
+		reader, err := az.bucket.NewReader(ctx, az.getObjectNameString(b), nil)
+		if err != nil {
+			return 0, err
+		}
+
+		defer reader.Close() //nolint:errcheck
+
+		throttled, err := az.downloadThrottler.AddReader(reader)
+		if err != nil {
+			return nil, err
+		}
+
+		b, err := ioutil.ReadAll(throttled)
+		if err != nil {
+			return nil, err
+		}
+
+		if len(b) != int(length) && length > 0 {
+			return nil, errors.Errorf("invalid length, got %v bytes, but expected %v", len(b), length)
+		}
+
+		if length == 0 {
+			return []byte{}, nil
+		}
+
+		return b, nil
+	}
+
+	v, err := exponentialBackoff(fmt.Sprintf("GetBlob(%q,%v,%v)", b, offset, length), attempt)
+	if err != nil {
+		return nil, translateError(err)
+	}
+
+	return v.([]byte), nil
+}
+
+func exponentialBackoff(desc string, att retry.AttemptFunc) (interface{}, error) {
+	return retry.WithExponentialBackoff(desc, att, isRetriableError)
+}
+
+func isRetriableError(err error) bool {
+	if me, ok := err.(azblob.ResponseError); ok {
+		if me.Response() == nil {
+			return false
+		}
+		// retry on server errors, not on client errors
+		return me.Response().StatusCode >= 500
+	}
+
+	return false
+}
+
+func translateError(err error) error {
+	switch gcerrors.Code(err) {
+	case gcerrors.NotFound:
+		return blob.ErrBlobNotFound
+
+	// TODO: Handle other possible cases
+	default:
+		return err
+	}
+	return err
+}
+
+func (az *azStorage) PutBlob(ctx context.Context, b blob.ID, data []byte) error {
+	// TODO: Perform chunked upload
```

Comment:
@PrasadG193 - We may not need to do this since Kopia already breaks data into chunks. @julio-lopez - can you confirm?

Reply:
Chunked uploads are not only not needed, but more importantly not desirable for the Kopia use case.
In order to meet the desired consistency guarantees, `PutBlob` must be done with a single `PUT` operation. Not sure what the semantics are for Azure blob, but other blob stores have weaker consistency guarantees for multi-part uploads.

Reply:
Ok. Will do the upload in single PUT then